### PR TITLE
Implement `?` macro repetition

### PR DESCRIPTION
**Pull request by mark-i-m**
See rust-lang/rfcs#2298 (with disposition merge)

### CODE REVIEW

**Comment by kennytm** [comment]
Marking as blocked by #47732, #47603, and of course rust-lang/rfcs#2298. I don't think this can be easily reviewed with all those formatting change 😂

**Comment by mark-i-m** [comment]
@kennytm I 100% agree. Thanks :)

**Comment by kennytm** [comment]
@mark-i-m Please put the new `?` syntax behind a feature gate since this is an unstable feature. You may refer to `:vis` (search for `macro_vis_matcher`) to know how this works.

**Comment by mark-i-m** [comment]
@kennytm Done.

**Comment by mark-i-m** [comment]
This doesn't have a tracking issue yet, so when that happens, I should update the feature gate...

**Comment by mark-i-m** [comment]
Also the unstable book...

File: src/libsyntax/ext/tt/quoted.rs
Comment (nikomatsakis) [inline src/libsyntax/ext/tt/quoted.rs]: Nit: can we add `assert_eq!(op, KleeneOp::ZeroOrOne);` here? I find it helps catch mistakes when later somebody messes with the arms accidentally....

File: src/test/compile-fail/macro-at-most-once-rep-ambig.rs
Comment (nikomatsakis) [inline src/test/compile-fail/macro-at-most-once-rep-ambig.rs]: I always appreciate a comment here explaining what this test is trying to do. Example:

---

Test that in `$(a)?+`, the `?` is interpreted as a separator, but not for `$(a)?`. This test focuses on error cases.

File: src/test/run-pass/macro-at-most-once-rep.rs
Comment (nikomatsakis) [inline src/test/run-pass/macro-at-most-once-rep.rs]: what about `$(a)?*`? we should test that too, no?

File: src/test/compile-fail/macro-at-most-once-rep-ambig.rs
Comment (nikomatsakis) [inline src/test/compile-fail/macro-at-most-once-rep-ambig.rs]: what about `$(a)?*`? we should test that too, no?

File: src/test/run-pass/macro-at-most-once-rep.rs
Comment (nikomatsakis) [inline src/test/run-pass/macro-at-most-once-rep.rs]: comment =)

File: src/test/run-pass/macro-at-most-once-rep.rs
Comment (nikomatsakis) [inline src/test/run-pass/macro-at-most-once-rep.rs]: these tests would be better if they had things in the body that made use of `a`

File: src/test/compile-fail/issue-39388.rs
Comment (nikomatsakis) [inline src/test/compile-fail/issue-39388.rs]: Please move this to a `ui` test instead of hard-coding spans and things (i.e., the `14:22`)

File: src/test/run-pass/macro-at-most-once-rep.rs
Comment (nikomatsakis) [inline src/test/run-pass/macro-at-most-once-rep.rs]: wait-- so `$(a),?` is *accepted* but the `,` is just a no-op? seems like it should be an error

File: src/test/ui/feature-gate-macro_at_most_once_rep.rs
Comment (nikomatsakis) [inline src/test/ui/feature-gate-macro_at_most_once_rep.rs]: This comment is wrong =)

**Review by nikomatsakis — CHANGES_REQUESTED** [review]
Good start, nice work! Left various comments, will re-review once addressed.

File: src/test/compile-fail/issue-39388.rs
Comment (mark-i-m) [reply src/test/compile-fail/issue-39388.rs]: Oh, I intended to remove the line/col numbers but forgot... Should I still move to a `ui` test if I remove them?

**Comment by mark-i-m** [comment]
@nikomatsakis 

> wait-- so $(a),? is accepted but the , is just a no-op? seems like it should be an error

How would you feel about a lint instead? I would like to keep the separator simply for consistency with `+` and `*` even though having a separator on "at most once" is silly.

**Comment by mark-i-m** [comment]
@nikomatsakis I think I addressed your comments. You may want to wait on a re-review until #47732, #47603 are merged, as they will make the diffs easier to read...

**Comment by mark-i-m** [comment]
> until #47732, #47603 are merged

They have merged now :tada: ... I will rebase shortly

File: src/test/compile-fail/issue-39388.rs
Comment (nikomatsakis) [reply src/test/compile-fail/issue-39388.rs]: Doesn't matter so much. I believe `ui` tests are generally "the future" but this test can stay a compile-fail test.

**Comment by nikomatsakis** [comment]
> How would you feel about a lint instead? I would like to keep the separator simply for consistency with + and * even though having a separator on "at most once" is silly.

Maybe we should bring this up in the RFC thread and get others' opinions. I don't have a *very* strong opinion, but it seems like an error would be better to me.

**Comment by mark-i-m** [comment]
@nikomatsakis I just rebased this PR... The diff should be _much_ more readable now. The only thing remaining is that when there is a tracking issue, I need to update the feature gate.

**Review by nikomatsakis — APPROVED** [review]

**Comment by nikomatsakis** [comment]
@mark-i-m would you rather merge this now, and update the tracking issue in a follow-up PR? Seems fine to me. r=me if so.

**Comment by kennytm** [comment]
I'd prefer just file a tracking issue now.

**Comment by mark-i-m** [comment]
TBH, it does feel a bit weird to have a tracking issue and merged implementation before the RFC is formally accepted, but if everyone agrees, that's fine with me 👍

**Comment by nikomatsakis** [comment]
Tracking issue created: https://github.com/rust-lang/rust/issues/48075

@mark-i-m can you update the PR to reference it?

**Comment by mark-i-m** [comment]
Updated

**Comment by CAD97** [comment]
Failure is due to added `(see tracking issue)`:

```
[00:53:32] ---- [ui] ui/feature-gate-macro_at_most_once_rep.rs stdout ----
[00:53:32] 	diff of stderr:
[00:53:32] 
[00:53:32] -	error[E0658]: Using the `?` macro Kleene operator for "at most one" repetition is unstable
[00:53:32] +	error[E0658]: Using the `?` macro Kleene operator for "at most one" repetition is unstable (see issue #48075)
```

**Comment by mark-i-m** [comment]
Thanks! Fixed.

**Comment by mark-i-m** [comment]
hmm... tidy is now complaining that that line is too long 🤕

**Pull request merged**
